src/cudamatrix/cu-kernels.cu

```diff
@@ -748,6 +748,27 @@ static void _add_mat_mat_elements(Real *data, const Real *srcA_data, const Real
     }
 }
 
+template<typename Real>
+__global__
+static void _add_mat_smat(Real *data, int stride, Real alpha, const Real *srcA_data, const int srcA_stride, bool transA, const RowElement<Real>* srcB_data, const int srcB_stride, Real beta) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x; // row index of output matrix
+  int32_cuda j = blockIdx.y; // column index of output matrix
+  int32_cuda tgt_index = i * stride; // actual index of data
+  int32_cuda srcA_index = -1, srcB_index = -1, sparse_index = -1;
+  Real tgt_value = beta * data[tgt_index];
+  for (int k = 0; k < srcB_stride; k++) {
+    srcB_index = j * srcB_stride + k;
+    sparse_index = srcB_data[srcB_index].column;
+    if (sparse_index == -1)
```

Comment:
Conditional statements are bad for pipelining.  It's better if you move the conditional out of the loop over k.